code/modules/projectiles/autofire.dm

```diff
@@ -1,70 +1,130 @@
 /**
 
-Full-auto firing by Kmc2000
-This system lets you spray and pray with guns when dragging the mouse.
+Full-auto firing by Kmc2000 (Version 2)
+So it turns out the way I wrote this before was absolutely abysmal (who would've guessed!) this should be a little bit more clean and less buggy <3
 
-*/
-
-/obj/item/gun/proc/can_fire_at(atom/target, mob/user)
-	if(!target)
-		return FALSE
+Fields:
 
-	if(target == user) //so we can't shoot ourselves with autofire
-		return FALSE
-
-	if(user.stat != CONSCIOUS) //No firing in softcrit
-		return FALSE
+full_auto - Set this if you want a full auto setting on your gun. This REPLACES semi-auto, but don't fret! You can still use semi auto mode when on full auto, just don't hold your mouse down :)
+It's highly recommended that you DO NOT use this for special case guns like the beam rifle, or guns that override mouseDown, as this will cause issues.
 
-	if(user.get_active_held_item() != src)
-		return FALSE
-
-	if(istype(target, /obj/screen))
-		return FALSE
-
-	if(target in user.contents) //can't shoot stuff inside us.
-		return FALSE
+Everything else should be handled for you. Good luck soldier.
 
-	if(user.a_intent == INTENT_HARM) //melee attacks are handled by attackby, not autofire
-		return FALSE
-
-	if(isliving(user))//Check if the user can use the gun, if the user isn't alive(turrets) assume it can.
-		var/mob/living/L = user
-		if(!can_trigger_gun(L))
-			return FALSE
-		if(clumsy_check)//If theyre a clown, they can't autofire.
-			if (HAS_TRAIT(user, TRAIT_CLUMSY))
-				return FALSE
+*/
 
-	if(!can_shoot()) //Just because you can pull the trigger doesn't mean it can shoot. (*CLICK*)
+#define COMSIG_AUTOFIRE_END "stop_autofiring"
+
+/obj/item/gun
+	var/full_auto = FALSE //Set this if your gun uses full auto. ONLY guns that go brr should use this. Not pistols!
+	var/datum/component/full_auto/autofire_component = null //Repeated calls to getComponent aren't really ideal. So we'll take the memory hit instead.
+
+/obj/item/gun/vv_edit_var(var_name, var_value)
+	. = ..()
+	switch(var_name)
+		if(NAMEOF(src, full_auto))
+			canMouseDown = var_value //Necessary to fully apply the component.
+			if(var_value) //Admin editing the autofire to true on a gun, let's help them out.
+				if(!autofire_component)
+					autofire_component = AddComponent(/datum/component/full_auto)
+					if(fire_rate)
+						autofire_component.default_fire_delay = (10 / fire_rate)
+					return ..()
```

Comment:
You run ..() twice

Reply:
True, none of this is explicitly necessary and I just did it for clarity